src/software/new_geom/spline2d.h

```diff
@@ -0,0 +1,118 @@
+#include "software/new_geom/point.h"
+#include "software/new_geom/polynomial2d.h"
+
+// Pre-declaration so we can make this a friend class of SplineSegment2d
+class Spline2d;
+
+/**
+ * A segment of a spline, defined by a 2d polynomial, and the start and end values that
+ * define the section of the 2d polynomial that is part of the spline
+ */
+class SplineSegment2d
+{
+   public:
+    SplineSegment2d() = delete;
+
+    double getStartVal() const;
+
+    double getEndVal() const;
+
+    Polynomial2d getPolynomial() const;
+
+   private:
+    friend class Spline2d;
+
+    /**
+     * Construct a SplineSegment2d
+     *
+     * This is private since we only want to be able to create one inside the Spline2d
+     * class (a friend class of this one)
+     *
+     * @param start_val The first value at which to evaluate the polynomial that lies on
+     *                  the spline
+     * @param end_val The last value at which to evaluate the polynomial that lies on
+     *                the spline
+     * @param polynomial The polynomial, part of which is on the spline
+     */
+    SplineSegment2d(double start_val, double end_val, Polynomial2d polynomial);
+
+    const double start_val;
+    const double end_val;
+    const Polynomial2d polynomial;
+};
+
+/**
+ * A 2D spline, consistent of a series of connected 2D polynomials.
+ *
+ * Important terms:
+ *      "knot" - A point on the spline where two consecutive polynomials meet
+ */
+class Spline2d
```

Comment:
Since `SplineSegment2d` is the same for linear or higher order splines, `getValueAt`, `getKnots`, `getNumKnots`, etc. is the same no matter what order Spline, so a derived class of Spline2d only needs to implement a different constructor. Can we implement these functions in Spline2d?

Reply:
I'm going to hold off on this one if that's OK. I definitely noticed the same thing, but the next thing I'm going to do is implement a cubic spline, which should make it very clear what can actually be shared.

Reply:
OK just take note of it somewhere

Reply:
Ok, added it as an acceptance criteria under https://github.com/UBC-Thunderbots/Software/issues/1382, since this will really all be done together.